### PR TITLE
Adding AT Modem binding

### PR DESCRIPTION
**Pull request by Ellerbach**
## Description

Adding AT Modem binding to support SMS, Call, HTTP(s), MQTT(s), File Storage for SIM800 (except MQTT) and SIM7080, SIM7070 and SIM7090.

## Motivation and Context

- From the #837, the idea came to build something more generic allowing to use different models. So credits to @Bengkel as well for starting this journey
- From the Eclo Solutions [implementation of the SIM800H](https://github.com/Eclo/nanoFramework.SIM800H), ideas on how to reuse HttpClient, and how to split the code with independent chunks of features thru properties. Still the SIM800H implementation from Eclo is more complete than this one for this specific model.
- Have generic implementations for AT standards like network connection, SMS and Calls that can work with any modem right out of the box with minimal efforts.
- Have an easy way to support HttpClient and MqttClient to make it easy to reuse code across different implementations thru Wifi, Ethernet and AT Modem.

## How Has This Been Tested?

* Intensively tested, the sample gives most of the tests
* Thru the Azure IoT library with the adjustment of the IMqttClient interface (samples to be added once this PR is merged)

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [x] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [x] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [x] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

File: devices/AtModem/Http/Http/Uri.cs [hunk 1]
```diff
@@ -0,0 +1,1210 @@
...
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Constructs Uri from string and enumeration that tell what is the type of Uri.
+        /// </summary>
+        /// <param name="uriString">String to construct Uri from</param>
+        /// <param name="kind">Type of Uri to construct</param>
+        /// <exception cref="ArgumentException">The scheme specified in the URI formed by combining <paramref name="baseUri"/> and <paramref name="relativeUri"/> is not valid.</exception>
```
Comment (josesimoes) [inline devices/AtModem/Http/Http/Uri.cs:428]: This method doesn't have any `ArgumentException` rather several `FormatException`...

File: devices/AtModem/Http/Http/HttpHeaders.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+// Portions Copyright (c) Microsoft Corporation.  All rights reserved.
+// See LICENSE file in the project root for full license information.
+//
+
+using System.Diagnostics;
+
+namespace System.Net.Http.Headers
+{
+    /// <summary>
+    /// Key/value pairs of headers. The value is either a raw <see cref="string"/> or a <see cref="HttpHeaders.HeaderStoreItemInfo"/>.
```
Comment (josesimoes) [inline devices/AtModem/Http/Http/HttpHeaders.cs:12]: We don't have `HeaderStoreItemInfo`

File: devices/AtModem/Http/Http/HttpClient.cs [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+        /// <value>
+        /// One of the values defined in the <see cref="Security.SslProtocols"/> enumeration. Default value is <see cref="SslProtocols.Tls12"/>.
+        /// </value>
+        /// <remarks>
+        /// This property is an extension from the full .NET required by nanoFramework.
+        /// </remarks>
+        public SslProtocols SslProtocols { get; set; } = SslProtocols.Tls12;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="HttpClient"/> class using a <see cref="HttpClientHandler"/> that is disposed when this instance is disposed.
```
Comment (josesimoes) [inline devices/AtModem/Http/Http/HttpClient.cs:97]: We don't have `HttpClientHandler`

File: devices/AtModem/Http/Http/HttpVersion.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+// Portions Copyright (c) Microsoft Corporation.  All rights reserved.
+// See LICENSE file in the project root for full license information.
+//
+
+namespace System.Net
+{
+
+    /// <summary>
+    /// Defines the HTTP version numbers that are supported by the
+    /// <see cref="HttpWebRequest"/> and
```
Comment (josesimoes) [inline devices/AtModem/Http/Http/HttpVersion.cs:12]: Can't corss ref to `HttpWebRequest` nor `HttpWebResponse`

File: devices/AtModem/AtModem.nfproj [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    <Reference Include="System.IO.Ports">
+      <HintPath>packages\nanoFramework.System.IO.Ports.1.1.60\lib\System.IO.Ports.dll</HintPath>
+    </Reference>
+    <Reference Include="System.IO.Streams">
+      <HintPath>packages\nanoFramework.System.IO.Streams.1.1.38\lib\System.IO.Streams.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Math">
+      <HintPath>packages\nanoFramework.System.Math.1.5.29\lib\System.Math.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Net">
```
Comment (josesimoes) [inline devices/AtModem/AtModem.nfproj:259]: Sys.Net shouldn't be referenced... right?

File: devices/AtModem/DTOs/BatteryStatus.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        /// <summary>
+        /// Gets the battery charge status.
+        /// </summary>
+        public BatteryChargeStatus Status { get; }
+
+        /// <summary>
+        /// Gets the charge level of the battery.
+        /// </summary>
+        public Ratio ChargeLevel { get; }
```
Comment (josesimoes) [inline devices/AtModem/DTOs/BatteryStatus.cs:21]: Do we need to depend & ref 2 UnitsNet libraries _just_ to format the charge ration and battery voltage?

File: devices/AtModem/AtModem.nfproj (hunk 5 above)
Comment (Ellerbach) [reply devices/AtModem/AtModem.nfproj]: correct, good catch! Removed!

File: devices/AtModem/DTOs/BatteryStatus.cs (hunk 6 above)
Comment (Ellerbach) [reply devices/AtModem/DTOs/BatteryStatus.cs:21]: Yes, that's our rule all up for all bindings. I removed the ratio from the signal class, as it's not really a ratio. I left it with the battery.

File: devices/AtModem/Http/Http/HttpVersion.cs (hunk 4 above)
Comment (Ellerbach) [reply devices/AtModem/Http/Http/HttpVersion.cs:12]: cleared. Good cat h!

File: devices/AtModem/Http/Http/HttpHeaders.cs (hunk 2 above)
Comment (Ellerbach) [reply devices/AtModem/Http/Http/HttpHeaders.cs]: I've cleaning as much as I could find. Those files are copy/paste for convenience. But agree, comments should be as close as the reality.

File: devices/AtModem/Http/Http/HttpKnownHeaderNames.cs [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+    /// <summary>
+    /// TBD
+    /// </summary>
+    internal class HttpKnownHeaderValues
+    {
+        /// <summary>TBD</summary>
+        public const string close = "close";
+    }
+
+    /*
```
Comment (josesimoes) [inline devices/AtModem/Http/Http/HttpKnownHeaderNames.cs:180]: Guess this is a leftover from the copy&paste... 😉

File: devices/AtModem/Http/Http/HttpKnownHeaderNames.cs [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+        // Force an an init, before we use them
+        private static HttpVerb[] m_knownVerbs;
+        static KnownVerbs()
+        {
+            m_knownVerbs = new HttpVerb[5];
+            m_knownVerbs[0] = new HttpVerb("GET", false, true, false, false);
+            m_knownVerbs[1] = new HttpVerb("POST", true, false, false, false);
+            m_knownVerbs[2] = new HttpVerb("HEAD", false, true, false, true);
+            m_knownVerbs[3] = new HttpVerb("PUT", true, false, false, false);
+            /*
```
Comment (josesimoes) [inline devices/AtModem/Http/Http/HttpKnownHeaderNames.cs:269]: Same here

File: devices/AtModem/Http/Http/ValidationHelper.cs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            {
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        // Not used in this context, keeping those in case.
```
Comment (josesimoes) [inline devices/AtModem/Http/Http/ValidationHelper.cs:67]: Maybe remove commented code?

**Review by josesimoes — APPROVED** [review]
Brilliant! Thank you for a LOT the amount of work you've put into this! 👏🏻 💯

**Pull request merged**